lib/Sema/CSFix.cpp

```diff
@@ -255,17 +255,51 @@ ContextualMismatch *ContextualMismatch::create(ConstraintSystem &cs, Type lhs,
   return new (cs.getAllocator()) ContextualMismatch(cs, lhs, rhs, locator);
 }
 
-bool AllowTupleTypeMismatch::diagnose(bool asNote) const {
-  auto failure = TupleContextualFailure(
-      getConstraintSystem(), getFromType(), getToType(), getLocator());
+bool AllowTupleTypeMismatch::coalesceAndDiagnose(
+    ArrayRef<ConstraintFix *> fixes, bool asNote) const {
```

Comment:
We'll actually need those fixes when we want to keep track of which tuple elements didn't match. I'm going to add that back now so it's ready for when we add the note for which elements didn't match. 

The reason this is necessary is because when we create the `AllowTupleTypeMismatch` fix, we only know the type of the specific element that didn't match - we don't know the full tuple type to put it into context. We're computing that in this method, and we want to coalesce the fixes together because we don't want to report the same error for each part of the tuple that didn't match. The reason why the fixes aren't being used here is because the fixes only contain the isolated element type that didn't match, but I'm changing it to also include the element index (the locator has this info before we strip it off). So, if we're matching tuple types `(Int, Float)` and `(Int, Int)`, when we create the fix we only know that `Int` and `Float` didn't match. There is no overarching tuple diagnostic in this case. 

It does seem strange to me that we're never using these the types that we record in the case of an element mismatch, though. I considered creating separate fixes for element mismatch, label mismatch, and artity mismatch, but tuples can have all three errors and with the way we're reporting it, it makes sense to have one error message for the contextual mismatch that contains the full tuple types and notes saying which parts structurally are wrong (i.e. `tuples have a different type for element #1`, `tuples have a different number of elements`, etc)

Reply:
This way of dealing with structural failures seems a bit too complicated to me... 

Maybe we should start with simpler approach and consider `element` to be `label + type` that way either one or both are going to be the same failure. I think it might make sense to consider aggregating element mismatches in `matchTupleTypes` where we have access to both top level tuples and also do matching of their elements. It seems like we don't really need to record particulars about how types mismatched, just the fact that they did (so if  https://github.com/apple/swift/blob/master/lib/Sema/CSSimplify.cpp#L1218 fails we record the position just like we do in fix for generic arguments) which means that we can potentially keep this all contained in one place...

The only problem with that approach might be when one side is generic parameter but we can also record anything `unsolved` as potential failure and skip it in `repairFailures` for `TupleElement` if there is a top level fix for that already. Do you know if any examples like that?

Reply:
This is a problem when we have any kind of type variable that we're matching in the tuples:
```swift
func foo(_ x: (Int, Double)) {}

foo((1.0, 2.0))
```
`matchTupleTypes` "succeeds" because we're comparing `($T1, $T2)` and `(Int, Double)`. In this case, `matchTupleTypes` is only called during CSGen.

Reply:
Hm, yeah this is really unfortunately and we can't delay destructuring because some of these types could be contextual... Maybe we should expend the locator with `tuple` element e.g. `tuple -> tuple element #N`?

Reply:
It seems like we need to introduce a generalized `ApplyArgToParam` element for structural comparisons like that to make it work.

Reply:
Yeah, that's exactly the information we would need. I think we're going to hit the same problem with `matchFunctionTypes` as we have here